resolve gene ambiguity reporting exception during profile import

sheridancbio opened the pull request:
# What? Why?
### Background
During import of profile datafiles, the entrez_gene_id for each record is resolved. Records with an explicit entrez_gene_id use the explicit entrez_gene_id but if not, we accept a hugo_gene_symbol field which is converted to entrez_gene_id. If an exact match to a hugo_gene_symbol in the database gene table is not found, gene_aliases are considered. When we encounter hugo_gene_symbols which map to multiple entrez_gene_ids, the importer will reject / skip these rows in the profile datafile usually. Two exceptions are allowed : one for RPPA profiles, and one for cases which involve entrez_gene_ids which are of type "miRNA". For the miRNA exception, we attempt to import the current datafile record for every matched entrez_gene_id of type "miRNA". The reason for doing this is presence of the specially constructed microRNA combination form (precursor/mature) gene table records created with negative value entrez_gene_ids by the deployment script ImportMicroRNAIDs.java  --- because CNA profiles for precursor gene symbol aliases (such as "hsa-mir-100") will match combination form gene record "MIR-100/100" and also combination form gene record "MIR-100/100*" which combines the precursor with either the 3p or the 5p arm mature form) and the CNA profile is applicable to both of these entrez_gene_id.

Another factor is that once a profile record has been imported for any entrez_gene_id, further attempts to import profile records for the same entrez_gene_id (within the same data file) will be rejected / skipped as duplicated records.

### The bug and the fix
The bug occurred in the logging code which reported skipped (due to duplication or ambiguity) records. The logging code attempted to distinguish duplicated microRNA records (which are expected to normally match multiple entrez_gene_ids) from ambiguous records of other types (protein coding genes mainly), but the logic would not handle cases where a gene alias matches a mixture of microRNA entrez_gene_ids plus a single/unique protein coding entrez_gene_id. The fix here is combine the logging of ambiguity problems involving non-microRNA genes with the logging of ambiguity problems involving a mixture of microRNA and non-microRNA genes into a single clause.

- cases where a hugo gene symbol was ambiguous (matching both microRNA and non-microRNA) and duplicated in a data profile file threw exceptions.
- now we consolidated the logging of these ambiguities and do not throw exceptions
- additional logging for cases where microRNA import was attempted but failed (versus no microRNAs being involved)

Co-authored-by: Manda Wilson <1458628+mandawilson@users.noreply.github.com>
Co-authored-by: Avery Wang <18199796+averyniceday@users.noreply.github.com>
Co-authored-by: Angelica Ochoa <aochoa4230@gmail.com>

# Checks
- [ ] Runs on Heroku.
- [ ] Follows [7 rules of great commit messages](http://chris.beams.io/posts/git-commit/). For most PRs a single commit should suffice, in some cases multiple topical commits can be useful. During review it is ok to see tiny commits (e.g. Fix reviewer comments), but right before the code gets merged to master or rc branch, any such commits should be squashed since they are useless to the other developers. Definitely avoid [merge commits, use rebase instead.](http://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/)
- [ ] Follows the [Google Style Guide](https://github.com/google/styleguide).
- [ ] If this is a feature, the PR is to rc. If this is a bug fix, the PR is to master.

# Any screenshots or GIFs?
If this is a new visual feature please add a before/after screenshot or gif
here with e.g. [GifGrabber](http://www.gifgrabber.com/).

pieterlukasse commented on core/src/main/java/org/mskcc/cbio/portal/scripts/ImportTabDelimData.java:
```diff
@@ -436,17 +440,19 @@ else if (!entrez.matches("[0-9]+")) {
+                                    // this case :
+                                    //      - the hugo gene symbol was ambiguous (matched multiple entrez-gene-ids)
+                                    //      - at least one of the entrez-gene-ids was not a microRNA (and this is not an rppaProfile)
+                                    //      - all of the matched microRNA ids (if any) failed to be imported (presumably already imported on a prior line)
```
I like these comments 👍 . Should we perhaps add the respective code / refactor the existing code to reflect these rules explicitly?

n1zea144 replied:
I would say its worth the effort if we plan to keep this file, but our current plan is to start cleaning up the core later this year, which will be followed up by resurrecting the spring-batch based importer.

sheridancbio commented:
rescinding request for review to @averyniceday by request

sheridancbio commented:
@pieterlukasse I think what we will do to address your points is to add additional comments in the code explaining the logic of this method with examples. Writing unit tests or refactoring the code is outside of the scope of this effort. The comments around the throwing of the runtime exception were added when we did not expect cases where a gene alias would refer both to protein coding genes in the gene table and also miRNA genes in the gene table. The logging of duplications and ambiguities which require skipping lines in the profile data file were basically categorized into 2 cases : case#1 : if there were more than one non-microRNA gene involved, we reported that we skipped the record due to being unable to resolve the gene symbol into a single entrez_gene_id. case#2 : if all the genes involved were microRNA, we reported that we were unable to import the current (miRNA) profile line from the datafile. What we did not expect was a gene alias matching exactly one protein-coding gene, and also several microRNA genes. The reporting code threw the exception in this case. If you consider that matching 2 protein-coding genes and also several microRNA genes did *not* generate the exception ... and that no protein-coding genes are even attempted to be imported if any microRNA genes are involved, maybe it makes a good argument for not throwing an exception in the case that exactly one protein coding gene is involved plus one or more microRNA (which had previously been imported to).

sheridancbio commented:
@pieterlukasse I have updated the documentation above this function to summarize the logical flow and explain the microRNA considerations. I've also added sample input / output cases which could function as unit test definitions. For simpler examination, I've generated the javadoc for this class (including private members) and attached it.
[ImportTabDelimData.pdf](https://github.com/cBioPortal/cbioportal/files/3299134/ImportTabDelimData.pdf)

sheridancbio commented:
I've fixed the line in the documentation which printed out as "4. after looping for several involved genes, ..."; now it should show up with a bullet point under the two prior lines ...
1.
2.
• after looping for several involved genes, ...

sheridancbio commented:
I've fixed the incorrect reference to table "genetic_alterations" and replaced it with "genetic_alteration"

sheridancbio commented:
[ImportTabDelimData.pdf](https://github.com/cBioPortal/cbioportal/files/3299181/ImportTabDelimData.pdf)

New version of javadoc

pieterlukasse commented:
thanks @sheridancbio. I like the extended documentation and test cases. These are a good basis for unit tests indeed. 
One part of the code that I still don't understand is: what happens when `discretizedCnaProfile` is `true` but `genes.size() > 1`? Will there be no CNA events added in `cnaEventsToAdd`? The code seems to contradict your documentation at
> This allows the visualization of both CNA event profiles for the microRNA precursor with expression profiles for the microRNA mature form.`

sheridancbio commented:
@pieterlukasse When genes.size() > 1, that takes precedence in the code; discretizedCnaProfile will not even be considered in that case.  If genes.size() > 1, and none of the matched genes are of type "miRNA", then no imports will be attempted for he current line. (One exception : rppaProfiles, which have special rules). If genes.size() > 1 and there *are* some genes of type miRNA, then there will be an attempt to import the profile on the current line for each gene of type miRNA. Failures will be logged, but if any of the miRNA genes are successfully used to import the current line profile line, the overall call to parseLine() is considered a success.

sheridancbio commented:
I rebased and made a small language adjustment (based on feedback from @mandawilson).

If the documentation seems to adequately explain the use cases and (complicated) code logic, we would like to move forward with merging this PR and importing a couple of studies which need this bugfix to allow import.

sheridancbio commented:
@pieterlukasse I see your point about discretizedCnaProfile .. you are right; for microRNA genes I think there will be no CNA Events to add to the sample_cna_event table. Maybe this was an error when we refactored this code adding the separate sample_cna_event table to hold events rather than storing the comma separated profile of events in genetic_alteration. I'll discuss this with the pipelines team. Fixing that error can be done in a separate PR ... but thank you for pointing it out.

sheridancbio commented:
Dismissing change request from @pieterlukasse to allow merge (under the belief that the documentation additions have sufficiently clarified the logic and use cases). I will create an issue for the sample_cna_event table non-use for microRNAs and we will fix that error in a follow up PR.

pieterlukasse commented:
@sheridancbio thanks for following up with an extra fix. One request: next time, please don't dismiss a change request before giving me enough time to reply. I prefer to do this myself. Please consider the time difference (I am not in Boston yet :stuck_out_tongue_winking_eye: ).
For the record: I think this PR does not introduce a new issue and is a step forward since the code better reflects a situation that is now known to be expected. I still think there should be a set of unit tests that can work as regression tests. This will make the future refactoring easier, since one can then easily check the new implementation against the existing tests.
